python/src/scipp/compat/dict.py

```diff
@@ -0,0 +1,227 @@
+    returned.
+    If the input keys contain both `'dims'` and `'values'`, as Variable is
+    returned.
+    Otherwise, a Dataset is returned.
+
+    :param dict_obj: A python dict to be converted to a scipp object.
+    :type dict_obj: dict
+    :return: A scipp Variable, DataArray or Dataset.
+    :rtype: Variable, DataArray, or Dataset
+    """
+    keys_as_set = set(dict_obj.keys())
+    if ({"coords", "data"}.issubset(keys_as_set)
+            or {"coords", "unaligned"}.issubset(keys_as_set)):
+        # Case of a DataArray-like dict (most-likely)
+        return _dict_to_data_array(dict_obj)
+    elif (keys_as_set.issubset(
+        {"dims", "values", "variances", "unit", "dtype", "shape"})
+          or keys_as_set.issubset(
+              {"value", "variance", "unit", "dtype", "shape", "dims"})):
+        # Case of a Variable-like dict (most-likely)
+        return _dict_to_variable(dict_obj)
+    else:
+        # Case of a Dataset-like dict
+        out = sc.Dataset()
+        for key, item in dict_obj.items():
+            out[key] = _dict_to_data_array(item)
+        return out
+
+
+def _dict_to_variable(d):
+    """
+    Convert a python dict to a scipp Variable.
+    """
+    # The Variable constructor does not accept both `shape` and `values`. If
+    # `values` is present, remove `shape` from the list. Also remove `dims` in
+    # the case of a 0D variable.
+    keylist = list(d.keys())
+    if "value" in keylist:
+        if "shape" in keylist:
+            keylist.remove("shape")
+        if "dims" in keylist:
+            keylist.remove("dims")
+    if "values" in keylist and "shape" in keylist:
+        keylist.remove("shape")
+    out = {}
+
+    is_event_data = False
+    if "dtype" in keylist:
+        is_event_data = str(d["dtype"]).startswith("event_list_float")
+
+    # TODO: maybe this constructor would be worth adding to the scipp module
+    # itself?
+    if is_event_data:
+        if "shape" not in d:
+            shp = d["values"].shape
+        else:
+            shp = d["shape"]
+
+        var = sc.Variable(dims=d["dims"],
+                          shape=shp,
+                          dtype=getattr(sc.dtype, str(d["dtype"])))
+
+        ndim = len(d["dims"])
+        indices = tuple()
+        for i in range(ndim):
+            indices += range(shp[i]),
+        # Now construct all indices combinations using itertools
+        for ind in product(*indices):
+            # And for each indices combination, slice the original data
+            vslice = var
+            aslice = d["values"]
+            for i in range(ndim):
+                vslice = vslice[d["dims"][i], ind[i]]
+                aslice = aslice[ind[i]]
+            vslice.values = aslice
+        return var
+
+    else:
+        for key in keylist:
+            if key == "dtype" and isinstance(d[key], str):
+                out[key] = getattr(sc.dtype, d[key])
+            else:
+                out[key] = d[key]
+        return sc.Variable(**out)
```

Comment:
I think this last line is almost everything this function should do:
- `Variable` constructor does accept keyword arguments, so passing a dict is possible.
- Only thing we may want to do here is to convert from a generic Python/numpy dict to a dict of things that scipp understands, e.g., string unit to scipp units, since we might not want that in the constructor.